atef/widgets/happi.py

```diff
@@ -0,0 +1,369 @@
+    device_selection_group: QtWidgets.QGroupBox
+    layout_by_name: QtWidgets.QHBoxLayout
+    radio_by_category: QtWidgets.QRadioButton
+    radio_by_name: QtWidgets.QRadioButton
+    list_or_tree_frame: QtWidgets.QFrame
+    button_refresh: QtWidgets.QPushButton
+    happi_tree_view: HappiDeviceTreeView
+    happi_list_view: HappiDeviceListView
+    _tree_updated: bool
+    _tree_current_category: str
+    _search_thread: Optional[ThreadWorker]
+
+    def __init__(
+        self,
+        parent: Optional[QWidget] = None,
+        *,
+        client: Optional[happi.Client] = None,
+    ):
+        super().__init__(parent=parent)
+        self._client = None
+        self._tree_current_category = "beamline"
+        self._search_thread = None
+        self._tree_has_data = False
+        self._setup_ui()
+        # Set the client at the end, as this may trigger an update:
+        self.client = client
+
+    def _setup_ui(self):
+        self.happi_tree_view.setVisible(False)
+        self._setup_tree_view()
+
+        self.button_refresh.clicked.connect(self.refresh_happi)
+        self.list_or_tree_frame.layout().insertWidget(0, self.happi_list_view)
+
+        self.radio_by_name.clicked.connect(self._select_device_widget)
+        self.radio_by_category.clicked.connect(self._select_device_widget)
+        self.combo_by_category.currentTextChanged.connect(self._category_changed)
+        self.button_refresh.clicked.emit()
+
+        def list_selection_changed(
+            selected: QtCore.QItemSelection, deselected: QtCore.QItemSelection
+        ):
+            self.happi_items_selected.emit([idx.data() for idx in selected.indexes()])
+
+        self.happi_list_view.selectionModel().selectionChanged.connect(
+            list_selection_changed
+        )
+
+    def _setup_tree_view(self):
+        """Set up the happi_tree_view if not already configured."""
+        view = self.happi_tree_view
+        view.groups = [
+            self.combo_by_category.itemText(idx)
+            for idx in range(self.combo_by_category.count())
+        ]
+        self.list_or_tree_frame.layout().insertWidget(0, view)
+
+        def tree_selection_changed(
+            selected: QtCore.QItemSelection, deselected: QtCore.QItemSelection
+        ):
+            items = [
+                idx.data() for idx in selected.indexes()
+                if idx.parent().data() is not None  # skip top-level items
+            ]
+            self.happi_items_selected.emit(items)
+
+        view.selectionModel().selectionChanged.connect(
+            tree_selection_changed
+        )
+
+        self.happi_tree_view = view
+
+    @property
+    def selected_device_widget(self) -> Union[HappiDeviceListView, HappiDeviceTreeView]:
+        """The selected device widget - either the list or tree view."""
+        if self.radio_by_name.isChecked():
+            return self.happi_list_view
+
+        return self.happi_tree_view
+
+    @QtCore.Slot(str)
```

Comment:
What does an explicit slot decorator do in pyqt?

Reply:
It's generally unnecessary and more of a personal habit I had when making a signal to annotate the slot. The primary reason would be for disambiguation when the signal itself is overloaded (i.e., multiple call signatures) - in C++ this would be automatically determined by the compiler, but in Python types are just a suggestion after all. According to PyQt5's [documentation](https://www.riverbankcomputing.com/static/Docs/PyQt5/signals_slots.html):

> Although PyQt5 allows any Python callable to be used as a slot when connecting signals, it is sometimes necessary to explicitly mark a Python method as being a Qt slot and to provide a C++ signature for it. PyQt5 provides the [pyqtSlot()](https://www.riverbankcomputing.com/static/Docs/PyQt5/api/qtcore/qtcore-module.html#pyqtSlot) function decorator to do this.
> ... (clipped)
> Connecting a signal to a decorated Python method also has the advantage of reducing the amount of memory used and is slightly faster.
> ...
> It is also possible to chain the decorators in order to define a Python method several times with different signatures. For example: